Implement Couchbase Collection Datanode and Add Unit Tests

snehaamujri opened the pull request:
Issue [#1862](https://github.com/Avaiga/taipy/issues/1862)

- Added  to handle data interactions with Couchbase collections.
- Introduced  for unit testing the functionality of the Couchbase Collection Datanode.
- Ensured the implementation aligns with existing data node standards and includes necessary validation.

snehaamujri committed:
Implement Couchbase Collection Datanode and Add Unit Tests

- Added  to handle data interactions with Couchbase collections.
- Introduced  for unit testing the functionality of the Couchbase Collection Datanode.
- Ensured the implementation aligns with existing data node standards and includes necessary validation.

snehaamujri committed:
Merge branch 'develop' into couchbase-datanode

jrobinAV commented on taipy/core/data/couchbase_collection_datanode.py:
```diff
@@ -0,0 +1,132 @@
+    def _read(self):
+        """Read all documents from the Couchbase collection."""
+        try:
+            return [doc.content_as[dict] for doc in documents]
```
What is documents here?

jrobinAV commented on taipy/core/data/couchbase_collection_datanode.py:
```diff
@@ -0,0 +1,132 @@
+        )
+
+        # Create a Couchbase connection using the provided properties.
+        try:
+            self.cluster = Cluster(
+                f"couchbase://{properties.get(self.__DB_HOST_KEY, self.__DB_HOST_DEFAULT)}",
+                ClusterOptions(
+                    PasswordAuthenticator(
+                        properties.get(self.__DB_USERNAME_KEY, ""),
+                        properties.get(self.__DB_PASSWORD_KEY, "")
+                    )
+                )
+            )
+            bucket = self.cluster.bucket(properties.get(self.__DB_NAME_KEY))
+            self.collection = bucket.collection(properties.get(self.__COLLECTION_KEY))
```
Would help the review if you could you provide some couchbase documentation link explaining this API. Thx in advance

jrobinAV commented on taipy/core/data/couchbase_collection_datanode.py:
```diff
@@ -0,0 +1,132 @@
+            return [doc.content_as[dict] for doc in documents]
+        except Exception as e:
+             print(f"An error occurred: {e}")
+             return []
```
TODO: We have to handle this case.

snehaamujri replied:
Thank you for the review and the detailed feedback! I appreciate the direction, and I'm eager to work on the improvements.

-> For document structure, I will define a Python class or object to structure the documents, ensuring consistency in what 
      is accepted.
-> I will proceed with implementing the _write, _append, and filter methods. I understand these are essential for CRUD 
     operations.
-> For testing, I will create a Mock class to simulate Couchbase interactions, allowing us to test without needing an 
     actual Couchbase server.

jrobinAV reviewed (commented):
That's promising. Good starting point. I believe the overall idea is correct.

Now we need to go into the real implementation. There are some missing points. For example: 
- what kind of documents are accepted? Do we need to define a Python object to structure the document?
- We are missing the _write, _append, and filter methods implementations
- For testing, we should have a Mock class acting like a CouchBase server without having to instantiate a real server
- etc...

snehaamujri committed:
Merge branch 'develop' into couchbase-datanode

snehaamujri committed:
Implemented changes in couchbase_collection_datanode.py, added mock class for testing, and updated documentation.

snehaamujri committed:
Implemented changes in couchbase_collection_datanode.py, added mock class for testing, updated documentation, and included all required features.

snehaamujri commented:
Hey @jrobinAV 

Resolved outstanding issues in couchbase_collection_datanode.py: 
- Created Couchbase connection in the try block using provided properties
- Implemented _write, _append, and filter methods 
- Added mock class in mock_couchbase.py for unit testing without real server
- Updated Couchbase API documentation reference
- Enhanced the document structure and Python object handling for clarity
- Provided additional documentation for new methods and classes

trgiangdo commented on taipy/core/data/couchbase_collection_datanode.py:
```diff
@@ -0,0 +1,183 @@
+                f"couchbase://{properties.get(self.__DB_HOST_KEY, self.__DB_HOST_DEFAULT)}",
+                ClusterOptions(
+                    PsswordAuthenticator(
+                        properties.get(self.__DB_USERNAME_KEY, ""),
+                        properties.get(self.__DB_PASSWORD_KEY, "")
```
I'm quite sure that Couchbase requires username and password to connect.

If the `properties` doesn't contain the `username` and `password`, we should raise an error before this, meaning that we should add the `self.__DB_USERNAME_KEY` and `self.__DB_PASSWORD_KEY` to the `self._REQUIRED_PROPERTIES`

trgiangdo commented on taipy/core/data/couchbase_collection_datanode.py:
```diff
@@ -0,0 +1,183 @@
+                    )
+                )
+            )
+            bucket = self.cluster.bucket(properties.get(self.__DB_NAME_KEY))
```
```suggestion
            bucket = self.cluster.bucket(properties.get(self.__BUCKET_KEY))
```

The constant should be renamed to match what it actually presents

trgiangdo commented on taipy/core/data/couchbase_collection_datanode.py:
```diff
@@ -0,0 +1,183 @@
+
+    def _write(self,data : Union[Dict, List[Dict]]):
+           """Write Documents to the Couchbase collection."""
+        try:
+            if isinstance(data, dict):
+                self.collection.upsert(data['id'],data)
+            elif isinstance(data, list):
+                for item in data:
+                   self.collection.upsert(item['id'], item)
+       except CouchbaseException as e:
+            print(f"An error occurred while writing:{e}")
```
```suggestion
        try:
            if isinstance(data, dict):
                self.collection.upsert(data['id'],data)
            elif isinstance(data, list):
                for item in data:
                   self.collection.upsert(item['id'], item)
        except CouchbaseException as e:
            print(f"An error occurred while writing:{e}")
```

The `except` is misalign

trgiangdo commented on taipy/core/data/couchbase_collection_datanode.py:
```diff
@@ -0,0 +1,183 @@
+from .data_node_id import DataNodeId, Edit
+
+
+class CouchbaseDocument:
+    """Class to define the structure of documents stored in Couchbase."""
+
+    def __init__(self, field1: str, field2: int, **kwargs):
+        self.field1 = field1  # Example field of type string
+        self.field2 = field2  # Example field of type integer
+        # Additional fields can be added dynamically
+        for key, value in kwargs.items():
+            setattr(self, key, value)
```
Where is this class being used?
What is it used for?

trgiangdo commented on taipy/core/data/couchbase_collection_datanode.py:
```diff
@@ -0,0 +1,183 @@
+           """Write Documents to the Couchbase collection."""
+        try:
+            if isinstance(data, dict):
+                self.collection.upsert(data['id'],data)
+            elif isinstance(data, list):
+                for item in data:
+                   self.collection.upsert(item['id'], item)
```
What is the motivation for using `upsert` here?
What should happens if the document doesn't have an `id`?

trgiangdo commented on tests/mocks/mock_couchbase.py:
```diff
@@ -0,0 +1,32 @@
+from couchbase_collection_datanode import CouchBaseCollectionDataNode  # Make sure this import is correct
+
+
+class TestCouchBaseCollectionDataNode(unittest.TestCase):
```
Can we add a `test_read` to this class?

trgiangdo commented on tests/core/data/mock_couchbase.py:
```diff
@@ -0,0 +1,59 @@
+class MockCouchbaseCollection:
+    def __init__(self):
+        self.documents = {}
+
+    def insert(self, doc_id, document):
+        if doc_id in self.documents:
+            raise Exception(f"Document with ID {doc_id} already exists.")
+        self.documents[doc_id] = document
+
+    def upsert(self, doc_id, document):
+        self.documents[doc_id] = document
+
+    def get(self, doc_id):
+        if doc_id not in self.documents:
+            raise Exception(f"Document with ID {doc_id} not found.")
+        return self.documents[doc_id]
+
+    def remove(self, doc_id):
+        if doc_id not in self.documents:
+            raise Exception(f"Document with ID {doc_id} not found.")
+        del self.documents[doc_id]
```
As my understanding, the MockCouchbaseCollection is essentially a dictionary right?

trgiangdo commented on tests/mocks/mock_couchbase.py:
```diff
@@ -0,0 +1,32 @@
+    def test_append_document(self):
+        doc_id = "doc2"
+        document = {"name": "Jane", "hobbies": []}
+        self.data_node._write(doc_id, document)  # Assuming _write is implemented correctly
```
Does this work?

I'm asking because the signature of the `CouchBaseCollectionDataNode._write()` method only requires 1 argument.
Here you are providing 2.

trgiangdo commented on tests/mocks/mock_couchbase.py:
```diff
@@ -0,0 +1,32 @@
+        self.data_node._write(doc_id, document)  # Assuming _write is implemented correctly
+
+        additional_data = ["reading", "hiking"]
+        self.data_node._append(doc_id, additional_data)  # Assuming _append is implemented correctly
```
Similarly, does `_append()` with 2 arguments work here?

trgiangdo reviewed (commented):
Hello @snehaamujri, thank you for your contribution.

I have some questions about the implementation.

jrobinAV commented:
@snehaamujri Your first version is promising but not completed yet. Have you had a chance to see our reviews? Do you have any questions?

snehaamujri commented:
> @snehaamujri Your first version is promising but not completed yet. Have you had a chance to see our reviews? Do you have any questions?

snehaamujri commented:
Hey @jrobinAV,
Thank you for your feedback! Yes, I have reviewed the comments, and I’m working on addressing the issues raised. I would appreciate any additional insights you have as I make the necessary changes. Please let me know if there are specific areas you’d like to collaborate on and i need help and if you have suggestions for improvements.

snehaamujri committed:
Merge branch 'develop' into couchbase-datanode

jrobinAV commented:
@snehaamujri 
Is there any news about resolving the comments we made on this PR?

jrobinAV commented:
Without any answer, we have to close the PR. Thank you for your understanding.
